powerquery-docs/data-privacy-firewall.md

```diff
@@ -240,7 +240,17 @@ Now we perform the static grouping. This maintains separation between partitions
 
 ![Post static-grouping firewall partitions.](media/data-privacy-firewall/firewall-steps-pane-3.png)
 
-Now we enter the dynamic phase. In this phase, the above static partitions are evaluated. Partitions that don't access any data sources are trimmed. Partitions are then grouped to create source partitions that are as large as possible. However, in this sample scenario, all the remaining partitions access data sources, and there isn't any further grouping that can be done. The partitions in our sample thus won't change during this phase.
+Now we enter the dynamic phase. In this phase, the above static partitions are evaluated. Partitions that don't access any data sources are trimmed. Partitions are then grouped to create source partitions that are as large as possible.
+
+### Why the dynamic phase?
```

Comment:
Your hypothesized answer is unfortunately incorrect.

As far as I can remember, the dynamic phase is done to minimize buffering due to partition boundaries, since that can drastically affect perf when running with the firewall on.

It uses what we call "analysis" to gather info about the data sources accessed. "Analysis" is the mechanism responsible for retrieving the 1k row previews we cache behind the scenes in PQ.

Reply:
I appreciate your response, @ehrenMSFT! Thank you. I am trying to figure the firewall out better, so much appreciated.

So, to check whether I'm following: :-)

* After the static analysis phase, the dynamic phase runs. Its analysis fetch performs in essence a `Table.FirstN(partition, 1000)` on each of the various partitions to learn about what is going on in the respective partition. 
* IIRC, the analysis fetch is also needed to determine which privacy level is associated with each data source resource (but maybe I'm recalling wrong here).

Something like that? 

If the entire partition is being evaluated during the analysis fetch, how do you keep cross-source data leakage from taking place during these fetches, as a partition could contain code that accesses more than one source?

Also, I'm curious: Do you know why 1,000 rows are fetched, instead of just 1?

Reply:
The analysis fetch isn't needed to determine privacy levels, but just to determine what data sources (if any) each partition accesses.

Cross-source data leakage is prevented by the firewall itself, which will block a partition from accessing a data source if its privacy level is incompatible with the sources previously accessed in the partition.

The 1,000 rows comes from the analysis we perform for caching the user-visible previews. The firewall just uses the same system.

Reply:
Thanks, @ehrenMSFT!

I think I'm missing a couple pieces here. :-)

> The analysis fetch isn't needed to determine privacy levels, but just to determine what data sources (if any) each partition accesses.

Would it be fair to say that the analysis fetch is a prerequisite to determining privacy levels, because the data sources being accessed need to be identified before their privacy levels can be looked up? 

> Cross-source data leakage is prevented by the firewall itself, which will block a partition from accessing a data source if its privacy level is incompatible with the sources previously accessed in the partition.

Can you expound a little more on this? :-) Earlier in the article, one of the main points is that the firewall replaces references to other partitions with `Value.Firewall` function calls so that the firewall can control the flow of data across partition boundaries.

However, during the analysis ("top 1000") evaluations, it sounds like the smallest unit of evaluation is a single partition, with no `Value.Firewall` references mixed into it. A single partition could exist which references multiple data sources—but the firewall doesn't know what data sources are accessed until it evaluates the partition. 

If it is evaluating the entire partition as a unit to figure out its data sources (if any), how does it stop data from flowing between those sources since a) it doesn't know about its sources until after evaluation and b) it isn't inserting `Value.Firewall` between them?

Reply:
1. Yes.
2. During the analysis, the Value.Firewall calls are still in place.

Reply:
Thanks, @ehrenMSFT!

I still don't think I'm following all the way. :-)

Say this is my query:
````
section Section1;

shared SomeQuery = 
let 
  Result = Table.Join(Sql.Database("a", "a"), "KeyA", Sql.Database("b", "b"), "KeyB")
in 
  Result;
````

The static phase would put step `Result` all in one partition, so no `Value.Firewall` calls would be intermixed inside of it, correct? 

If so, when the dynamic phase runs and evaluates this partition, how would the firewall keep data from potentially flowing between the two SQL sources, since no `Value.Firewall` calls are in place between the data source function invocations and the join?

Reply:
- Yes, the Result step would be in a single partition.
- The firewall doesn't just let a partition access any data sources it wants to. When a data source is about to be accessed, the firewall checks the compatibility of the source's privacy level against that of the other data sources already accessed in that partition. Unlike the cross-partition case, the firewall doesn't buffer to prevent data leakage within a single partition. Instead, it uses a heavy hammer and simply blocks data sources with incompatible privacy levels from being accessed in the same partition.

Reply:
I appreciate your patient help, @ehrenMSFT! 

[Ben](https://github.com/MicrosoftDocs/powerquery-docs/pull/513#discussion_r1489679675)> Would it be fair to say that the analysis fetch is a prerequisite to determining privacy levels, because the data sources being accessed need to be identified before their privacy levels can be looked up?
[Ehren](https://github.com/MicrosoftDocs/powerquery-docs/pull/513#discussion_r1490142856)> Yes

Then from the last comment:
Ehren> The firewall doesn't just let a partition access any data sources it wants to. When a data source is about to be accessed, the firewall checks the compatibility of the source's privacy level against that of the other data sources already accessed in that partition. Unlike the cross-partition case, the firewall doesn't buffer to prevent data leakage within a single partition. Instead, it uses a heavy hammer and simply blocks data sources with incompatible privacy levels from being accessed in the same partition.

Both of these statements make sense independently, but—I must be missing something here—seem to me to create a circular dependency when put together. :-)

* A: The firewall won't allow a single partition to be evaluated if it contains sources with incompatible privacy levels. Instead, it terminates evaluation with an error. Makes sense by itself. 
* B: In order to determine each data source's privacy level, an analysis fetch of the partition is required. By definition, the applicable privacy levels haven't been determined when this fetch occurs, because the analysis fetch has to occur as part of figuring that out. Makes sense by itself.
* A + B: Putting the two together—So, when the firewall encounters the single partition example from above, it won't evaluate it if it contains multiple sources with incompatible privacy levels (A) but it needs to evaluate it as a prerequisite to determining whether the sources in it have incompatible privacy levels (B). Seems like A can't happen before B, yet also B can't safely happen without A.

Reply:
The apparent contradiction is resolved by the fact that A is enforced dynamically at runtime, during the analysis evaluation.

Reply:
You make it sound like it's somehow planned out and done in pre-defined phases, which isn't the case.

Instead, the firewall just "runs" the M in the partition. When a data source function gets called, the function itself calls into the host to get the credentials for the data source, and to ensure it has permission to access the source.

Reply:
Thanks much, @ehrenMSFT. I appreciate your patience in explaining. :-)